From 9bcc1b2f41fa20b2b21c8283a8c402798416eee7 Mon Sep 17 00:00:00 2001
From: Heinz Wrobel <Heinz.Wrobel@nxp.com>
Date: Tue, 8 Nov 2022 11:33:55 +0100
Subject: [PATCH 04/20] pseudo: Added support function for fnmatch.

This function will be needed in subsequent commits to enable changes
to execution path checks.

Signed-off-by: Heinz Wrobel <Heinz.Wrobel@nxp.com>
---
 pseudo_client.c | 49 +++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 49 insertions(+)

diff --git a/pseudo_client.c b/pseudo_client.c
index a03d6b1..0510097 100644
--- a/pseudo_client.c
+++ b/pseudo_client.c
@@ -26,6 +26,7 @@
 #include <fcntl.h>
 #include <pwd.h>
 #include <grp.h>
+#include <fnmatch.h>
 
 #ifdef PSEUDO_XATTRDB
 #include <sys/xattr.h>
@@ -1503,6 +1504,54 @@ pseudo_client_shutdown(int wait_on_socket) {
 	return 0;
 }
 
+/* This function matches a PATH like exception list to the paths we use.
+ * NOTE: It has a length limit on any individual pattern!
+ */
+static char *patternlistmatchespath(const char *patternlist, const char *path, char *resbuf, size_t resbuflen, int checkalias)
+{
+	char *foundpath = NULL, *res;
+
+	/* For backwards compatibility, no patternlist means that
+	 * anything matches. THis is different from an empty list!
+	 */
+	if (patternlist && path) {
+		char patternbuf[NAME_MAX];
+		const char *s = patternlist;
+		size_t i,j;
+
+		if (!resbuf) {
+			resbuf = &patternbuf[0];
+			resbuflen = sizeof(patternbuf);
+		}
+
+		while (*s) {
+			s += strspn(s, ":");
+			i = strcspn(s, ":");
+			if (i && i < resbuflen) {
+				strncpy(resbuf, s, i);
+				resbuf[i] = 0;
+				res = &resbuf[0];
+				if(checkalias) {
+					j = strcspn(resbuf, "=");
+					if (j) {
+						resbuf[j++] = '\0';
+						res = &resbuf[j];
+					}
+					else
+						res = NULL;
+				}
+				if(fnmatch(resbuf, path, 0) == 0) {
+					foundpath = res;
+					break;
+				}
+			}
+			s += i;
+		}
+	}
+
+	return foundpath;
+}
+
 static char *
 base_path(int dirfd, const char *path, int leave_last) {
 	char *basepath = 0;
-- 
2.34.1

